posterior_averaging/shared/fed_pa_schedule.py

```diff
@@ -0,0 +1,820 @@
+ClientUpdateDeltaFn = Callable[..., Collection[tf.Tensor]]
+ClientWeightFn = Callable[..., float]
+LRScheduleFn = Callable[[Union[int, tf.Tensor]], Union[tf.Tensor, float]]
+
+
+def _initialize_optimizer_vars(model: tff.learning.Model,
+                               optimizer: tf.keras.optimizers.Optimizer):
+  """Ensures variables holding the state of `optimizer` are created."""
+  delta = tf.nest.map_structure(tf.zeros_like, _get_weights(model).trainable)
+  model_weights = _get_weights(model)
+  grads_and_vars = tf.nest.map_structure(lambda x, v: (x, v), delta,
+                                         model_weights.trainable)
+  optimizer.apply_gradients(grads_and_vars, name='server_update')
+  assert optimizer.variables()
+
+
+def _get_weights(model: tff.learning.Model) -> tff.learning.ModelWeights:
+  return tff.learning.ModelWeights.from_model(model)
+
+
+@attr.s(eq=False, order=False, frozen=True)
+class DataPassOutput(object):
+  """Structure for outputs returned by a single data pass.
+
+  Fields:
+  - `loss`: A scalar tensor that represents the average loss during the pass.
+  - `num_examples`: A scalar tensor that represents the number of seen examples.
+  - `model_weights_trainable`: A dictionary of model's trainable variables that
+      contain the weights obtained at the end of the data pass.
+  - `model_weights_trainable_sample`: A dictionary of model's trainable
+      variables that contain a candidate posterior sample.
+  """
+  loss = attr.ib()
+  num_examples = attr.ib()
+  model_weights_trainable = attr.ib()
+  model_weights_trainable_sample = attr.ib()
+
+
+@attr.s(eq=False, order=False, frozen=True)
+class DeltaUpdateOutput(object):
+  """Structure for outputs returned by delta update functions.
+
+  This structure contains the sufficient state of the dymanic program used by
+  PostAvg to update weight deltas given a new approximate posterior sample.
+  Each field described below is a dictionary of the same structure as model's
+  trainable variables.
+
+  Fields:
+  - `num_samples`: The number of approximate posterior samples so far.
+  - `weights_delta`: Updates to model's trainable variables.
+  - `weights_sample_mean`: The mean of the all posterior samples so far.
+      `n`-th sample from the mean of the previous `(n - 1)` samples.
+  - `dp_state`: The state of the dynamic programming (DP) represented by a
+      structure of `tf.TensorArray`s.
+  """
+  num_samples = attr.ib()
+  weights_delta = attr.ib()
+  weights_sample_mean = attr.ib(None)
+  dp_state = attr.ib(default=None)
+
+  @classmethod
+  def from_weights(cls, initial_weights, updated_weights, num_samples=0):
+    """Creates a delta update output from intial and updated weights."""
+    # Initialize updates.
+    weights_delta = tf.nest.map_structure(lambda a, b: a - b, updated_weights,
+                                          initial_weights)
+    # Initial state for dynamic programming updates.
+    vk_tas_init = tf.nest.map_structure(
```

Comment:
Can we give this a clearer name? Generally I'm against "variable" names, like "v", or time-varying variable names, like "vk".

Reply:
Ah, yes, I agree. The idea behind this notation was to match the equations that we provide in the paper, where `v_k` and `u_k` are particular intermediate vectors computed by our recursive algorithm. Let me think if I can improve the naming convention or maybe explain the naming convention beforehand in the docstring of the corresponding function.